### PR TITLE
yumpkg: don't use diff_attr when determining install/downgrade targets

### PR DESCRIPTION
**Pull request by terminalmage**
Doing so breaks epoch handling, and is unnecessary anyway since the diff_attr is only used for the return data.

Additionally, this tweaks the "attr" argument in both yumpkg and zypper's list_pkgs func so that it will accept a comma-separated list as well as a Python list, and makes a DRY tweak so that we're only returning and formatting the return data in one place in the function.

### CODE REVIEW

**Comment by terminalmage** [comment]
re-run lint

**Comment by brejoc** [comment]
Sorry, I couldn't make it today. I'll look at this on Monday morning.

**Review by brejoc — APPROVED** [review]
Looks good to me. Thanks, @terminalmage!

File: salt/modules/yumpkg.py [hunk 1]
```diff
@@ -650,42 +651,52 @@ def list_pkgs(versions_as_list=False, **kwargs):
...
-            __salt__['pkg_resource.add_pkg'](ret, pkginfo.name, all_attr)
-
-    for pkgname in ret:
-        ret[pkgname] = sorted(ret[pkgname], key=lambda d: d['version'])
-
-    __context__['pkg.list_pkgs'] = ret
+    contextkey = 'pkg.list_pkgs'
 
-    return __salt__['pkg_resource.format_pkg_list'](ret, versions_as_list, attr)
+    if contextkey not in __context__:
```
Comment (isbm) [inline salt/modules/yumpkg.py]: If I am reading this correctly, the `__context__[contextkey]` will be read only once per runtime and `rpm -qa` will _always_ return cached outdated values. Or I just missed something?..

File: salt/modules/zypper.py [hunk 2]
```diff
@@ -701,30 +702,36 @@ def list_pkgs(versions_as_list=False, **kwargs):
...
-        "%|EPOCH?{%{EPOCH}}:{}|_|-%{INSTALLTIME}\\n")]
-    ret = {}
-    for line in __salt__['cmd.run'](cmd, output_loglevel='trace', python_shell=False).splitlines():
-        name, pkgver, rel, arch, epoch, install_time = line.split('_|-')
-        install_date = datetime.datetime.utcfromtimestamp(int(install_time)).isoformat() + "Z"
-        install_date_time_t = int(install_time)
+    contextkey = 'pkg.list_pkgs'
+
+    if contextkey not in __context__:
+
```
Comment (isbm) [inline salt/modules/zypper.py]: Maybe this empty line is not needed.

**Review by isbm — COMMENTED** [review]
@terminalmage could you please double-check my concern? Thanks!

File: salt/modules/yumpkg.py (hunk 1 above)
Comment (terminalmage) [reply salt/modules/yumpkg.py]: This change is functionally equivalent to how we did things before, it just changes how we return data from the function. Any time packages are modified in Salt we clear that key from the `__context__` dunder (look for all the places we do `__context__.pop('pkg.list_pkgs', None)`).

**Review by rallytime — APPROVED** [review]

File: salt/modules/yumpkg.py (hunk 1 above)
Comment (isbm) [reply salt/modules/yumpkg.py]: Ah. That exactly the part I missed. Then it makes sense, thanks!

**Pull request merged**